daemon/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+/// Daemonizing will assert-fail if any additional threads have been started.
+///
+/// About all a program should do beforehand is to parse command-line options (needed to determine
+/// whether to detach) and initialize logging (because daemonization will emit log entries).
+#[derive(Clone, Debug, Default, Parser, PartialEq, Eq)]
+pub struct Daemonize {
+    /// Detach from foreground session
+    #[clap(long)]
+    pub detach: bool,
+
+    /// Create monitoring process
+    #[clap(long)]
+    pub monitor: bool,
+
+    /// Do not change directory to root
+    #[clap(long)]
+    pub no_chdir: bool,
+
+    /// Create pidfile
+    #[clap(long)]
+    pub pidfile: Option<PathBuf>,
+}
+
+impl Daemonize {
+    pub fn new() -> Daemonize {
+        Self::default()
+    }
+
+    /// Daemonizes the current process based on the configured options.  See the module comment for
+    /// full usage.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it forks and exits in the parent: any threads, other than
+    /// the calling thread, will be dead in the child.  Thus, it is only safe to call this function
+    /// while the program is single-threaded, and this function will assert-fail if additional
+    /// threads have been started.
+    pub unsafe fn start(self) -> (Daemonizing, Cleanup) {
+        Daemonizing::new(self)
+    }
+
+    pub unsafe fn run(self) -> Cleanup {
+        let (daemonizing, cleanup) = self.start();
+        daemonizing.finish();
+        cleanup
+    }
+}
+
+/// The current process, in the process of daemonizing.
+///
+/// Call `finish()` on this object to finish the daemonization process and notify the parent that
+/// it can exit successfully.  (If `finish()` is not called and the process does not exit, the
+/// parent will hang until one or the other occurs.)
+pub struct Daemonizing {
+    options: Daemonize,
+    notify_pipe: Option<File>
+}
+
+impl Daemonizing {
+    /// Completes the daemonization process:
+    ///
+    ///   - If we detached, changes the current directory to the root, unless that behavior is
+    ///     disabled.
+    ///
+    ///   - If we detached, closes the `stdin`, `stdout`, and `stderr` fds.  For safety, instead
+    ///     of leaving fds 0, 1, and 2 unpopulated, we replace them by `/dev/null`.
+    ///
+    ///   - Notifies the parent process that daemonization is complete.  This allows the parent
+    ///     process to exit successfully, indicating to the process that in turn
+    pub fn finish(mut self) {
+        if self.options.detach {
+            if !self.options.no_chdir {
+                drop(set_current_dir("/"));
+            }
+            close_standard_fds();
+        }
+        if let Some(ref mut pipe) = self.notify_pipe {
+            fork_notify_startup(pipe);
+        }
+    }
+
+    unsafe fn new(options: Daemonize) -> (Self, Cleanup) {
+        assert_single_threaded();
```

Comment:
I find it interesting that you do all this work in the constructor.
In C++/Java this is frowned upon - constructors should always terminate and initialize all the fields.
You'd leave this work for another method, e.g., run().

Reply:
I could call this method `start` or `run` if you want, since Rust doesn't distinguish constructors from other methods. Anyway, it's an implementation detail of this file since this isn't marked `pub`.